sonarqube/resource_sonarqube_portfolio.go

```diff
@@ -0,0 +1,364 @@
+	SelectionMode string   `json:"selectionMode"`
+	Branch        string   `json:"branch"`
+	Tags          []string `json:"tags"`
+	Regexp        string   `json:"regexp"`
+}
+
+// Returns the resource represented by this file.
+func resourceSonarqubePortfolio() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSonarqubePortfolioCreate,
+		Read:   resourceSonarqubePortfolioRead,
+		Update: resourceSonarqubePortfolioUpdate,
+		Delete: resourceSonarqubePortfolioDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceSonarqubePortfolioImport,
+		},
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"key": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"qualifier": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"visibility": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "public",
+				ForceNew: true, // I cant find an API endpoint for changing this, even though it's possible in the UI
+				ValidateFunc: func(val any, key string) (warns []string, errs []error) {
+					visibility := val.(string)
+					validOptions := []string{"public", "private"}
+					if !slices.Contains(validOptions, visibility) {
+						errs = append(errs, fmt.Errorf("Accepted values are public or private for key %q, got: %s", key, val))
+					}
+
+					return
+				},
+			},
+			"selection_mode": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "NONE",
+				ForceNew: false,
+				ValidateFunc: func(val any, key string) (warns []string, errs []error) {
+					selectionMode := val.(string)
+					validOptions := []string{"NONE", "MANUAL", "TAGS", "REGEXP", "REST"}
+					if !slices.Contains(validOptions, selectionMode) {
+						errs = append(errs, fmt.Errorf("Accepted values are NONE, MANUAL, TAGS, REGEXP or REST for key %q, got: %s", key, val))
+					}
+					return
+				},
+			},
+			"branch": { // Only active for TAGS, REGEXP and REST
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+			"tags": { // Only active for TAGS
+				Type:          schema.TypeList,
+				Optional:      true,
+				ForceNew:      false,
+				ConflictsWith: []string{"regexp"},
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"regexp": { // Only active for REGEXP
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      false,
+				ConflictsWith: []string{"tags"},
+			},
+
+			// TODO: MANUAL
+			// "selectedProjects": [],
+			// "projects": [],
+		},
+	}
+}
+
+// Validate the regexp and tag fields if the corresponding selection_mode is chosen
+func validatePortfolioResource(d *schema.ResourceData) error { // TODO: Doc this
```

Comment:
What is the `// TODO: Doc this` for?

Reply:
Old comment I forgot to remove. The line above it the doc 😄